### PR TITLE
docs: update wasm.md

### PR DESCRIPTION
**Pull request by wmitros**
The WASM UDF implementation has changed since the last time the docs were written. In particular, the Rust helper library has been released, and using it should be the recommended method. Some decisions that were only experimental at the start, were also "set in stone", so we should refer to them as such.

The docs also contain some code examples. This patch adds tests for these examples to make sure that they are not wrong and misleading.

### CODE REVIEW

File: docs/dev/wasm.md [hunk 1]
```diff
@@ -50,75 +50,69 @@ and passing for each parameter a 64-bit value, of which top 32 bits are its size
...
-If the function is specified not to accept NULL, all parameters are represented
-as in the description above.
+If the function is specified not to accept NULL, the parameter types natively supported by WASM are represented by their values (types shorter than 4 bytes are extended to 4 bytes),
+and all remaining types are represented as in the description above.
 
 If the function is specified to accept NULL, each parameter should be represented in WebAssembly by 
 size and pointer to its serialized form in memory, as in the paragraph above.
 
-the important distinction is that size equal to -1 (minus one)indicates that the value is NULL and should not be parsed.
+The important distinction is that size equal to -1 (minus one or i32::MAX) indicates that the value is NULL and should not be parsed.
```
Comment (avikivity) [inline docs/dev/wasm.md:64]: Is that not u32::MAX?

We shouldn't use rust terminology in wasm documentation, it's not limited to Rust, at least conceptually.

File: docs/dev/wasm.md [hunk 2]
```diff
@@ -50,75 +50,69 @@ and passing for each parameter a 64-bit value, of which top 32 bits are its size
...
-NOTE: ABI for return values is experimental and subject to change. It can (and should) be redesigned
-after implementing helper libraries for a few popular languages (including C++, C, Rust).
+NOTE: CQL syntax extensions, and new helper libraries may be deployed together with new ABI versions,
+the description below only refers to ABI versions 1 and 2.
 
-Natively supported types are returned as is. All the other types are returned via memory, similarly to the way
-they are passed as parameters: the wasm function should return a single 64-bit integer, representing the size and the offset of a serialized form of the returned value
+In `RETURNS NULL ON NULL INPUT` UDFs, natively supported types are returned as is. All the other types are returned via memory,
+similarly to the way they are passed as parameters: the wasm function should return a single 64-bit integer, representing
+the size and the offset of a serialized form of the returned value
```
Comment (avikivity) [inline docs/dev/wasm.md:73]: Let's add information about memory management. Who is responsible to free the memory, both for inputs and return values? How is memory freed?

Comment (wmitros) [reply docs/dev/wasm.md:73]: We already have this: https://github.com/scylladb/scylladb/blob/64102780fe570b53617594900eba1e229e537a3c/docs/dev/wasm.md?plain=1#L21-L29, is there anything to add?

File: docs/dev/wasm.md (hunk 1 above)
Comment (wmitros) [reply docs/dev/wasm.md:64]: Yes, it should be u32::MAX, or maybe, without using Rust terminology,  `0xffffffff`

File: docs/dev/wasm.md [hunk 3]
```diff
@@ -50,75 +50,69 @@ and passing for each parameter a 64-bit value, of which top 32 bits are its size
 }
 ```
 
-## Support for NULL values
+## Parameter values
```
Comment (nyh) [inline docs/dev/wasm.md:53]: Why did you change the title of this section? It seems like it is entirely about NULLs so the original title fit well, didn't it?

File: test/cql-pytest/test_wasm.py [hunk 4]
```diff
@@ -1107,3 +1107,97 @@ def test_counter(cql, test_keyspace, scylla_only):
             cql.execute(f"UPDATE {table} SET c = c + 1  WHERE p = 42;")
             cql.execute(f"UPDATE {table} SET c = c - 4  WHERE p = 42;")
             assert cql.execute(f"SELECT {ri_counter_name}(c) AS result FROM {table} WHERE p = 42").one().result == -1
+
+def test_docs_assemblyscript(cql, test_keyspace, table1, scylla_only):
```
Comment (nyh) [inline test/cql-pytest/test_wasm.py:1111]: Please add a comment how this test relates to the docs, and accurate information on how you compiled (?) the weird source code below, in case a future developer needs to recompile it after a change to the doc.

File: docs/dev/wasm.md (hunk 2 above)
Comment (avikivity) [reply docs/dev/wasm.md:73]: I forgot it. But yes, we should document the return value too.

One trick: suppose we have an aggregation function f(accumulator, arg) -> accumulator. The user is allowed to move the first argument into the return value (this moves the responsibility of freeing to the caller). We can then take advantage of this and *not* call _scylla_free on the return value; instead, we pass the return value back the next time we aggregate. This saves us the ffi overhead of extracting the accumulator and copying it back again, if we can reuse the function.

File: docs/dev/wasm.md (hunk 3 above)
Comment (wmitros) [reply docs/dev/wasm.md:53]: I changed it mostly because of the following section, which is called `Return values` and also mostly focuses on the behavior with NULLs, while this one only described the NULLs as UDF parameters. 
It seems to me like one of these 2 titles is wrong. Maybe we should have 2 subsections of the NULLs section, or maybe it would be better if we had a comprehensive section on NULLs, and put the parts that aren't NULL-specific somewhere else

File: test/cql-pytest/test_wasm.py (hunk 4 above)
Comment (wmitros) [reply test/cql-pytest/test_wasm.py:1112]: The UDF source code and compilation commands are described in the docs file.
I added a comment referring to this test in the docs, and a comment referring to the docs here (in addition to the suggestive name).

**Review by nyh — APPROVED** [review]

**Review by avikivity — APPROVED** [review]
